runtime-modules/project-token/src/tests/bonding_curve.rs

```diff
@@ -0,0 +1,1079 @@
+            IssueTokenFixture::default().execute_call().unwrap();
+
+            let result = UnbondFixture::default().execute_call();
+
+            assert_err!(result, Error::<Test>::NotInAmmState);
+        })
+}
+
+#[test]
+fn unbonding_order_fails_with_deadline_expired() {
+    let deadline_timestamp = 0u64;
+    let (user_account_id, user_balance) = (member!(2).1, joy!(5_000_000));
+    build_default_test_externalities_with_balances(vec![(user_account_id, user_balance)])
+        .execute_with(|| {
+            IssueTokenFixture::default().execute_call().unwrap();
+            ActivateAmmFixture::default().execute_call().unwrap();
+            BondFixture::default().execute_call().unwrap();
+
+            let result = UnbondFixture::default()
+                .with_deadline(deadline_timestamp)
+                .execute_call();
+
+            assert_err!(result, Error::<Test>::DeadlineExpired);
+        })
+}
+
+#[test]
+fn unbonding_order_failed_with_slippage_constraint_violated() {
+    let slippage_tolerance = (Permill::zero(), joy!(1_000_000_000));
+    let (user_account_id, user_balance) = (member!(2).1, joy!(5_000_000));
+    build_default_test_externalities_with_balances(vec![(user_account_id, user_balance)])
+        .execute_with(|| {
+            IssueTokenFixture::default().execute_call().unwrap();
+            ActivateAmmFixture::default().execute_call().unwrap();
+            BondFixture::default().execute_call().unwrap();
+
+            let result = UnbondFixture::default()
+                .with_slippage_tolerance(slippage_tolerance)
+                .execute_call();
+
+            assert_err!(result, Error::<Test>::SlippageToleranceExceeded);
+        })
+}
+
+#[test]
+fn amm_treasury_balance_correctly_decreased_during_unbonding() {
+    let token_id = token!(1);
+    let ((user_member_id, user_account_id), user_balance) = (member!(2), joy!(5_000_000));
+    build_default_test_externalities_with_balances(vec![(user_account_id, user_balance)])
+        .execute_with(|| {
+            IssueTokenFixture::default().execute_call().unwrap();
+            ActivateAmmFixture::default().execute_call().unwrap();
+            BondFixture::default()
+                .with_amount(DEFAULT_BONDING_AMOUNT)
+                .with_sender(user_account_id)
+                .with_member_id(user_member_id)
+                .execute_call()
+                .unwrap();
+            let amm_reserve_account = Token::module_bonding_curve_reserve_account(token_id);
+            let amm_reserve_pre = Balances::usable_balance(amm_reserve_account);
+
+            UnbondFixture::default()
+                .with_amount(DEFAULT_UNBONDING_AMOUNT)
+                .with_sender(user_account_id)
+                .with_member_id(user_member_id)
+                .execute_call()
+                .unwrap();
+
+            let amm_reserve_post = Balances::usable_balance(amm_reserve_account);
+            let correctly_computed_joy_amount = 100095;
```

Comment:
Please add a comment describing where this value comes from

Reply:
I added an helper function to make the computation clearer